File: services/api/app/schemas/cam_risk.py
# services/api/app/schemas/cam_risk.py
"""
CAM Risk Analytics Schema - Phase 18.0

Models for job risk reporting and timeline tracking.
Enables post-simulation risk analysis with severity tracking.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class RiskIssue(BaseModel):
    """Single risk issue from simulation."""
    index: int
    type: str
    severity: str  # info, low, medium, high, critical
    x: float
    y: float
    z: Optional[float] = None
    extra_time_s: Optional[float] = None
    note: Optional[str] = None
    meta: Dict[str, Any] = Field(default_factory=dict)


class RiskAnalytics(BaseModel):
    """Computed risk analytics from issues."""
    total_issues: int
    severity_counts: Dict[str, int]
    risk_score: float
    total_extra_time_s: float
    total_extra_time_human: str


class RiskReportIn(BaseModel):
    """Incoming risk report submission."""
    job_id: str = Field(..., description="User/job chosen ID, e.g., pipeline run ID or filename")
    pipeline_id: Optional[str] = Field(
        default=None, description="Optional pipeline definition/flow ID"
    )
    op_id: Optional[str] = Field(
        default=None, description="Which op produced these issues (e.g. 'rosette_sim')"
    )
    machine_profile_id: Optional[str] = None
    post_preset: Optional[str] = None

    design_source: Optional[str] = Field(
        default=None, description="e.g., 'dxf', 'svg', 'blueprint'"
    )
    design_path: Optional[str] = Field(
        default=None, description="Path to DXF/SVG/blueprint if known"
    )

    issues: List[RiskIssue] = Field(default_factory=list)
    analytics: RiskAnalytics


class RiskReportOut(BaseModel):
    """Stored risk report with metadata."""
    id: str
    created_at: datetime
    job_id: str
    pipeline_id: Optional[str]
    op_id: Optional[str]
    machine_profile_id: Optional[str]
    post_preset: Optional[str]
    design_source: Optional[str]
    design_path: Optional[str]
    issues: List[RiskIssue]
    analytics: RiskAnalytics

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class RiskReportSummary(BaseModel):
    """Lightweight summary for timeline browsing."""
    id: str
    created_at: datetime
    job_id: str
    pipeline_id: Optional[str]
    op_id: Optional[str]
    machine_profile_id: Optional[str]
    post_preset: Optional[str]
    
    # Analytics summary
    total_issues: int
    critical_count: int
    high_count: int
    medium_count: int
    low_count: int
    info_count: int
    risk_score: float
    total_extra_time_s: float

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }

File: packages/client/src/api/camRisk.ts
// client/src/api/camRisk.ts
/**
 * CAM Risk Analytics API Client - Phase 18.0
 * 
 * Client for job risk reporting and timeline browsing.
 */

export interface RiskIssue {
  index: number
  type: string
  severity: string
  x: number
  y: number
  z?: number | null
  extra_time_s?: number | null
  note?: string | null
  meta?: Record<string, any>
}

export interface RiskAnalytics {
  total_issues: number
  severity_counts: Record<string, number>
  risk_score: number
  total_extra_time_s: number
  total_extra_time_human: string
}

export interface RiskReportIn {
  job_id: string
  pipeline_id?: string | null
  op_id?: string | null
  machine_profile_id?: string | null
  post_preset?: string | null
  design_source?: string | null
  design_path?: string | null
  issues: RiskIssue[]
  analytics: RiskAnalytics
}

export interface RiskReportOut {
  id: string
  created_at: string
  job_id: string
  pipeline_id?: string | null
  op_id?: string | null
  machine_profile_id?: string | null
  post_preset?: string | null
  design_source?: string | null
  design_path?: string | null
  issues: RiskIssue[]
  analytics: RiskAnalytics
  meta?: Record<string, any>
}

// Alias for full stored reports (Phase 19.0)
export type RiskReportStored = RiskReportOut

export interface RiskReportSummary {
  id: string
  created_at: string
  job_id: string
  pipeline_id?: string | null
  op_id?: string | null
  machine_profile_id?: string | null
  post_preset?: string | null
  total_issues: number
  critical_count: number
  high_count: number
  medium_count: number
  low_count: number
  info_count: number
  risk_score: number
  total_extra_time_s: number
}

export interface CamRiskTimelineReport {
  id: string
  created_at: number
  lane: string
  job_id?: string | null
  preset?: string | null
  source?: string | null
  summary: Record<string, any>
}

export interface CamRiskTimelineQuery {
  lane?: string | null
  preset?: string | null
  limit?: number
  startTs?: number | null
  endTs?: number | null
}

/**
 * Submit a new CAM risk report.
 * Called after simulation to log risk analytics.
 */
export async function postRiskReport(report: RiskReportIn): Promise<RiskReportOut> {
  const response = await fetch('/api/cam/jobs/risk_report', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(report)
  })

  if (!response.ok) {
    const text = await response.text()
    throw new Error(text || `HTTP ${response.status}`)
  }

  return response.json()
}

/**
 * Get risk timeline for a specific job.
 * Returns all risk reports for the given job_id.
 */
export async function getJobRiskTimeline(
  jobId: string,
  limit: number = 50
): Promise<RiskReportOut[]> {
  const response = await fetch(`/api/cam/jobs/${jobId}/risk_timeline?limit=${limit}`)

  if (!response.ok) {
    const text = await response.text()
    throw new Error(text || `HTTP ${response.status}`)
  }

  return response.json()
}

/**
 * Get recent risk reports across all jobs.
 * Returns lightweight summaries for timeline browsing.
 */
export async function getRecentRiskReports(
  limit: number = 100
): Promise<RiskReportSummary[]> {
  const response = await fetch(`/api/cam/jobs/recent?limit=${limit}`)

  if (!response.ok) {
    const text = await response.text()
    throw new Error(text || `HTTP ${response.status}`)
  }

  return response.json()
}

export async function fetchCamRiskTimelineReports(
  query: CamRiskTimelineQuery = {},
): Promise<CamRiskTimelineReport[]> {
  const params = new URLSearchParams()
  if (query.lane) params.set('lane', query.lane)
  if (query.preset) params.set('preset', query.preset)
  if (query.limit) params.set('limit', String(query.limit))
  if (typeof query.startTs === 'number') params.set('start_ts', String(query.startTs))
  if (typeof query.endTs === 'number') params.set('end_ts', String(query.endTs))

  const qs = params.toString()
  const url = qs ? `/api/cam/risk/reports?${qs}` : '/api/cam/risk/reports'
  const response = await fetch(url)

  if (!response.ok) {
    const text = await response.text()
    throw new Error(text || `HTTP ${response.status}`)
  }

  return response.json()
}

// Legacy stubs for backward compatibility
export function buildRiskReportPayload(data: any): any {
  console.warn('buildRiskReportPayload: Legacy stub, use RiskReportIn interface directly')
  return data
}

export async function patchRiskNotes(_id: string, _notes: string): Promise<any> {
  console.warn('patchRiskNotes: Not implemented in Phase 18.0')
  return { success: false, message: 'Not implemented' }
}

export async function attachRiskBackplot(_id: string, _backplot: any): Promise<any> {
  console.warn('attachRiskBackplot: Not implemented in Phase 18.0')
  return { success: false, message: 'Not implemented' }
}

export interface RiskBackplotMoveOut {
  x?: number
  y?: number
  z?: number
  type?: string
}
